cmd/agent/app/builder.go

```diff
@@ -127,6 +129,21 @@ func (b *Builder) getReporter(primaryProxy CollectorProxy) reporter.Reporter {
 	return reporter.NewMultiReporter(rep...)
 }
 
+func (b *Builder) setExpvarOptions() {
+	for _, p := range b.Processors {
+		prefix := fmt.Sprintf(processorPrefixFmt, p.Model, p.Protocol)
+		if expvar.Get(prefix+suffixServerMaxPacketSize) == nil {
```

Comment:
This looks suspicious - there should be no other place in the code that would set this expvar, and if there were it would be good to panic because the data would be misleading.

I assume the reason for this check is because otherwise unit tests would panic? Which means it's masking the actual problem.

I am inclined to say that instead of writing to expvar directly (which uses global variables and thus is always problematic to test), we might want to use a dedicated metrics factory, which for unit tests can be instantiated to NullFactory, and in production (in main()s) it can be instantiated with expvar backing.

Reply:
That's right, some of the unit tests are panicing due to expvar rewriting that's why I check expvar variable before creating a new one.

Reply:
Would you be interested in following up on my suggestion above and using MetricsFactory abstraction instead of direct expvar access? Our jaeger-lib library provides an implementation of MetricsFactory backed by expvar.

Another option would be to use the same MetricsFactory that we already have, but introduce a convention that certain metrics namespaces, e.g. `internal` would be dispatched to a different implementation (expvar in this case), instead of being piped to the standard metrics exporter.

Reply:
Looks interesting, sure I'd give it a try but which option is preferable? For me it's still hard to get the whole picture.

Reply:
Having a dedicated MetricsFactory for internal parameters potentially requires more wiring, to pass it through all the way from main.  And it's a bit confusing since now we'd have two metric factories everywhere. 

So I would say the last option is most preferable, it relies on MetricsFactory that's already present almost everywhere.

Overall going through MetricsFactory API is preferable over direct expvar access because it allows more flexibility in where the parameters will be exposed in the end, e.g. if in the future we adopt OpenTelemetry SDK that supports z-pages, I think it would make more sense to expose this data via z-pages (can be a nicer UI) than via default expvar handler.

I think the only caveat with using `internal` namespace as a trigger  is that it will not always be the top-level namespace, because some components, such as UDP servers in the agent, are likely passed MetricsFactory that is already namespaced by the caller. But I think it's ok, we could have metrics like `ns1.ns2.internal.ns3.metricName` that would be forked to expvar.

In order to do this you first need to create a PR in jaeger-lib for a new dispatching MetricsFactory, e.g.
```
func NewNamespaceForkingMetricsFactory(
    forkNamespace string,
    forkFactory MetricsFactory,
    defaultFactory MetricsFactory,
) MetricsFactory
```

lmk if you need more details.

Reply:
So roughly in jaeger-lib we need to implement a separate factory like in [this example](https://github.com/jaegertracing/jaeger-lib/blob/master/metrics/multi/multi.go) and this factory will handle metrics with ```forkNamespace``` by ```forkFactory```?

Reply:
Yes! But it wouldn't be "multi" in the same way, i.e. when factory creates a counter, it would be just one counter from the defaultFactory. Basically it should always delegate to the default factory except for Namespace() function:

```go
func (f *Factory) Namespace(scope metrics.NSOptions) metrics.Factory {
    if scope.Name == f.forkNamespace {
        return f.forkFactory.Namespace(scope)
    }
    return &Factory({
        forkNamespace: f.forkNamespace,
        forkFactory: f.forkFactory.Namespace(scope),
        defaultFactory: f.defaultFactory.Namespace(scope),
    })
}
```